src/D2L.CodeStyle.Analyzers/Language/ImmutableGenericDeclarationAnalyzer.cs

```diff
@@ -67,21 +67,35 @@ KnownImmutableTypes knownImmutableTypes
 					continue;
 				}
 
-				var typeName = typeSymbol.GetFullTypeName();
-				if( typeName == "System.Nullable" ) {
-					var namedTypeSymbol = typeSymbol as INamedTypeSymbol;
-					if (namedTypeSymbol == default(INamedTypeSymbol)) {
-						continue;
-					}
-					typeSymbol = namedTypeSymbol.TypeArguments[ 0 ];
-				}
+				ValidateImmutability(
+					context,
+					typeSymbol,
+					knownImmutableTypes );
+			}
+		}
 
-				if( knownImmutableTypes.IsTypeKnownImmutable( typeSymbol ) ) {
-					continue;
+		private static void ValidateImmutability(
+			SyntaxNodeAnalysisContext context,
+			ITypeSymbol typeSymbol,
+			KnownImmutableTypes knownImmutableTypes
+		) {
+			if( ImmutableContainerMethods.IsAnImmutableContainerType( typeSymbol ) ) {
+
+				var namedTypeSymbol = typeSymbol as INamedTypeSymbol;
+				if( namedTypeSymbol == default( INamedTypeSymbol ) ) {
+					context.ReportDiagnostic( Diagnostic.Create(
+						Diagnostics.GenericArgumentTypeMustBeImmutable,
+						context.Node.GetLocation(),
+						messageArgs: new object[] { namedTypeSymbol.Name } ) );
 				}
 
-				ImmutabilityScope argumentScope = typeSymbol.GetImmutabilityScope();
-				if( argumentScope != ImmutabilityScope.SelfAndChildren ) {
+				foreach( var typeArgument in namedTypeSymbol.TypeArguments ) {
+
+					ValidateImmutability( context, typeArgument, knownImmutableTypes );
+				}
+			} else {
+				if( ( !knownImmutableTypes.IsTypeKnownImmutable( typeSymbol ) ) &&
+					( typeSymbol.GetImmutabilityScope() != ImmutabilityScope.SelfAndChildren ) ) {

```

Comment:
This looks like it will work, but it sucks that the standard way to say "is this random type immutable" is:
`typeSymbol.GetImmutabilityScope() != ImmutabilityScope.SelfAndChildren`
Maybe we can make that clearer later on.  It's not super clear as it is